feat: 채팅방 생성 및 게시물 작성자 자동 참여 기능 추가

Namgyu11 opened the pull request:
### 🚀 이 PR을 통해 해결하려는 문제
> 이 PR을 통해 해결하려는 문제를 적어주세요
- 게시물 작성자가 게시물을 생성할 때, 자동으로 해당 게시물에 대한 채팅방을 생성하고 작성자를 채팅방에 자동으로 참여시키는 기능이 없음.

### ✨ 이 PR에서 핵심적으로 변경된 사항
<!-- 문제를 해결하면서 주요하게 변경된 사항들을 적어 주세요 -->
> 문제를 해결하면서 주요하게 변경된 사항들을 적어 주세요 
- **채팅방 생성 및 작성자 자동 참여 기능 추가**:
  - `ChatRoomService`와 `ChatRoomMemberService`에서 게시물 생성 시 채팅방을 자동으로 생성하고, 게시물 작성자를 채팅방에 자동으로 참여시키는 로직을 추가
  - `AccompanyPostServiceImpl`에서 게시물 생성 후, `ChatRoomService.createChatRoom`을 호출하여 해당 게시물에 대한 채팅방을 생성하고 작성자를 참여

- **에러 처리 및 리팩토링**:
  - 채팅방 관련 에러 코드 `CHAT_ROOM_NOT_FOUND`와 `ALREADY_JOINED_CHAT_ROOM`을 추가하여 예외 상황을 명확하게 처리하도록 했습니다.
  - `GlobalExceptionHandler`에서 발생할 수 있는 일반적인 예외에 대해 일관된 HTTP 응답을 반환하도록 리팩토링했습니다.

### 핵심 변경 사항 외에 추가적으로 변경된 부분
<!-- 없으면 ‘없음’ 이라고 기재해 주세요 -->
> 없음

### 테스트
<!-- 본 변경사항이 테스트가 되었는지 기술해주세요 -->
- [ ] 테스트 코드 작성
- [ ] API 테스트 진행

Namgyu11 committed:
fix: 회원가입 요청 파라미터 설정 수정

- @RequestPart 어노테이션을 사용하여 SignUpDto 요청 파라미터를 명시했습니다.
- 이를 통해 회원가입 요청 시 올바른 파라미터가 전달되도록 수정했습니다.

Namgyu11 committed:
feat: 채팅방 관련 에러 코드 추가 및 예외 처리 리펙토링

- CHAT_ROOM_NOT_FOUND 및 ALREADY_JOINED_CHAT_ROOM 에러 코드를 추가했습니다.
- 회원가입 시 @RequestPart 어노테이션을 추가하여 요청 파라미터를 개선했습니다.
- GlobalExceptionHandler에서 예외 처리 시 ResponseEntity를 반환하도록 수정했습니다.

Namgyu11 committed:
feat: 채팅방 생성 및 게시물 작성자 자동 참여 기능 추가

- AccompanyPostServiceImpl에 ChatRoomService 의존성 추가 및 채팅방 생성 로직 추가
- ChatRoomServiceImpl에 채팅방 생성 메서드 추가
- ChatRoomMemberService 인터페이스 및 구현체 추가하여 사용자가 채팅방에 자동으로 참여하는 기능 구현
- 필요한 엔티티 및 리포지토리 수정 및 추가

49EHyeon42 commented on src/main/java/connectripbe/connectrip_be/chat/service/impl/ChatRoomMemberServiceImpl.java:
```diff
@@ -0,0 +1,63 @@
+
+        // 사용자가 존재하는지 확인
+        MemberEntity member = memberRepository.findById(memberId)
+                .orElseThrow(() -> new GlobalException(ErrorCode.USER_NOT_FOUND));
```
ErrorCode, USER_NOT_FOUND, NOT_FOUND_MEMBER 혼용 사용, 논의 필요

Namgyu11 replied:
확인 했습니당~

49EHyeon42 commented on src/main/java/connectripbe/connectrip_be/chat/service/impl/ChatRoomMemberServiceImpl.java:
```diff
@@ -0,0 +1,63 @@
+     *                          사용자가 이미 채팅방에 참여한 경우 발생하는 예외
+     */
+    @Override
+    public void jointChatRoom(Long chatRoomId, Long memberId) {
```
트랜잭션 설정 없음

Namgyu11 replied:
createAccompanyPost 메서드에서 이미 트랜잭션을 관리하고 있기 때문에, 
jointChatRoom 메서드에서는 별도의 트랜잭션을 설정하지 않아도 됩니다. 
jointChatRoom 메서드는 createAccompanyPost 트랜잭션 내에서 호출되므로, 데이터 일관성은 이미 보장됩니다. 
다만, 만약 이 메서드가 독립적으로 사용될 가능성이 있다면, 그때는 트랜잭션 설정을 추가하는 것으로 고려하겠습니다.

좋은 지적 감사합니다.😍

49EHyeon42 commented on src/main/java/connectripbe/connectrip_be/chat/service/impl/ChatRoomServiceImpl.java:
```diff
@@ -2,57 +2,96 @@
-                    .map(ChatRoomMemberResponse::fromEntity)
-                    .toList();
-      }
+    private static final Logger log = LoggerFactory.getLogger(ChatRoomServiceImpl.class);
```
`@slf4j`로 수정 필요

Namgyu11 replied:
수정 했습니다~

49EHyeon42 commented on src/main/java/connectripbe/connectrip_be/chat/service/impl/ChatRoomServiceImpl.java:
```diff
@@ -2,57 +2,96 @@
 
 @Service
 @RequiredArgsConstructor
 public class ChatRoomServiceImpl implements ChatRoomService {
```
트랜잭션 설정 없음

Namgyu11 replied:
https://github.com/100-hours-a-week/5-team-ark-connectrip-be/pull/63#discussion_r1730003209
감사합니다!

49EHyeon42 commented on src/main/java/connectripbe/connectrip_be/global/exception/GlobalExceptionHandler.java:
```diff
@@ -26,10 +26,11 @@ public ResponseEntity<ErrorResponse> handleCustomException(GlobalException e) {
     }
 
     @ExceptionHandler(Exception.class)
```
Exception.class 사용, 다른 커스텀 예외 설정 필요

Namgyu11 replied:
피드백 감사합니다!
Exception.class는 예상치 못한 예외 상황을 포괄적으로 처리하기 위해 사용했습니다. 
말씀해주신 대로 특정한 예외 상황에 대해 더 구체적인 커스텀 예외 클래스를 정의해보도록 하겠습니다~

49EHyeon42 commented on src/main/java/connectripbe/connectrip_be/global/exception/type/ErrorCode.java:
```diff
@@ -49,6 +49,9 @@ public enum ErrorCode {
     DUPLICATE_MEETING(HttpStatus.BAD_REQUEST, "이미 모임에 참여하셨습니다."),
 
+    // ChatRoom error
+    CHAT_ROOM_NOT_FOUND(HttpStatus.BAD_REQUEST, "해당 채팅방을 찾을 수 없습니다."),
```
찾지 못하는 경우, 404로 수정 필요

49EHyeon42 commented on src/main/java/connectripbe/connectrip_be/global/exception/type/ErrorCode.java:
```diff
@@ -49,6 +49,9 @@ public enum ErrorCode {
 
+    // ChatRoom error
+    CHAT_ROOM_NOT_FOUND(HttpStatus.BAD_REQUEST, "해당 채팅방을 찾을 수 없습니다."),
+    ALREADY_JOINED_CHAT_ROOM(HttpStatus.BAD_REQUEST, "이미 참여한 채팅방입니다."),
```
400도 좋지만, 충돌을 의미하는 409도 고려 필요

49EHyeon42 commented on src/main/java/connectripbe/connectrip_be/chat/repository/ChatRoomMemberRepository.java:
```diff
@@ -7,10 +7,9 @@
-      List<ChatRoomMemberEntity> findByMember_Email(String email);
-
-      List<ChatRoomMemberEntity> findByChatRoom_Id(Long chatRoomId);
+    List<ChatRoomMemberEntity> findByMember_Email(String email);
```
email이 아닌 id로 사용자를 조회할 수 있도록 수정 필요

Namgyu11 replied:
테스트 코드 때문에 잠시 남겨 놨는데, 빠른 시일 내에 삭제 하겠습니다~

Namgyu11 committed:
refactor: 불필요한 빈 줄 및 로거 코드 제거

- AccompanyPostServiceImpl 및 ChatRoomServiceImpl에서 불필요한 빈 줄을 제거했습니다.
- ChatRoomServiceImpl에서 사용되지 않는 로거 코드를 제거했습니다.

49EHyeon42 reviewed (approved):
LGTM 😎
